src/tests/end2end/BlendStateTests.cpp

```diff
@@ -0,0 +1,1055 @@
+        nxt::RenderPipeline testPipeline;
+        nxt::Texture renderTarget;
+        nxt::TextureView renderTargetView;
+        nxt::ShaderModule vsModule;
+        nxt::BindGroupLayout bindGroupLayout;
+        nxt::PipelineLayout pipelineLayout;
+};
+
+namespace {
+    // Add two colors and clamp
+    constexpr RGBA8 operator+(const RGBA8& col1, const RGBA8& col2) {
+        int r = static_cast<int>(col1.r) + static_cast<int>(col2.r);
+        int g = static_cast<int>(col1.g) + static_cast<int>(col2.g);
+        int b = static_cast<int>(col1.b) + static_cast<int>(col2.b);
+        int a = static_cast<int>(col1.a) + static_cast<int>(col2.a);
+        r = (r > 255 ? 255 : (r < 0 ? 0 : r));
+        g = (g > 255 ? 255 : (g < 0 ? 0 : g));
+        b = (b > 255 ? 255 : (b < 0 ? 0 : b));
+        a = (a > 255 ? 255 : (a < 0 ? 0 : a));
+
+        return RGBA8(static_cast<uint8_t>(r), static_cast<uint8_t>(g), static_cast<uint8_t>(b), static_cast<uint8_t>(a));
+    }
+
+    // Subtract two colors and clamp
+    constexpr RGBA8 operator-(const RGBA8& col1, const RGBA8& col2) {
+        int r = static_cast<int>(col1.r) - static_cast<int>(col2.r);
+        int g = static_cast<int>(col1.g) - static_cast<int>(col2.g);
+        int b = static_cast<int>(col1.b) - static_cast<int>(col2.b);
+        int a = static_cast<int>(col1.a) - static_cast<int>(col2.a);
+        r = (r > 255 ? 255 : (r < 0 ? 0 : r));
+        g = (g > 255 ? 255 : (g < 0 ? 0 : g));
+        b = (b > 255 ? 255 : (b < 0 ? 0 : b));
+        a = (a > 255 ? 255 : (a < 0 ? 0 : a));
+
+        return RGBA8(static_cast<uint8_t>(r), static_cast<uint8_t>(g), static_cast<uint8_t>(b), static_cast<uint8_t>(a));
+    }
+
+    // Get the component-wise minimum of two colors
+    RGBA8 min(const RGBA8& col1, const RGBA8& col2) {
+        return RGBA8(
+            std::min(col1.r, col2.r),
+            std::min(col1.g, col2.g),
+            std::min(col1.b, col2.b),
+            std::min(col1.a, col2.a)
+        );
+    }
+
+    // Get the component-wise maximum of two colors
+    RGBA8 max(const RGBA8& col1, const RGBA8& col2) {
+        return RGBA8(
+            std::max(col1.r, col2.r),
+            std::max(col1.g, col2.g),
+            std::max(col1.b, col2.b),
+            std::max(col1.a, col2.a)
+        );
+    }
+
+    // Blend two RGBA8 color values parameterized by the provided factors in the range [0.f, 1.f]
+    RGBA8 mix(const RGBA8& col1, const RGBA8& col2, std::array<float, 4> fac) {
+        float r = static_cast<float>(col1.r) * (1.f - fac[0]) + static_cast<float>(col2.r) * fac[0];
+        float g = static_cast<float>(col1.g) * (1.f - fac[1]) + static_cast<float>(col2.g) * fac[1];
+        float b = static_cast<float>(col1.b) * (1.f - fac[2]) + static_cast<float>(col2.b) * fac[2];
+        float a = static_cast<float>(col1.a) * (1.f - fac[3]) + static_cast<float>(col2.a) * fac[3];
+
+        // Get the integer locations of the resulting 4D RGBA8 hypercube after the blend
```

Comment:
Shouldn't round() be enough? I don't think the GPU does that euclidean distance computation.